File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* recurse(ListNode* prev, ListNode* curr){
        if (curr == nullptr){
            return prev;
        }
        ListNode* next = curr -> next;
        curr -> next = prev;
        return recurse(curr, next);
    }
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head -> next == nullptr){
            return head;
        }
        ListNode* prev = nullptr;
        return recurse(prev, head);
         
    }
};

File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        elif not head.next:
            return head
        prev = None
        current = head
        while current:
            next = current.next
            current.next = prev

            prev = current
            current = next
        return prev  
        

